opm/material/densead/EvaluationV.hpp

```diff
@@ -0,0 +1,602 @@
+/*!
+ * \file
+ *
+ * \brief Representation of an evaluation of a function and its derivatives w.r.t. a set
+ *        of variables in the localized OPM automatic differentiation (AD) framework.
+ */
+#ifndef OPM_DENSEAD_EVALUATION_V_HPP
+#define OPM_DENSEAD_EVALUATION_V_HPP
+
+#include "MathV.hpp"
+
+#include <opm/material/common/Valgrind.hpp>
+
+#include <dune/common/version.hh>
+
+#include <vector>
+#include <cmath>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <algorithm>
+
+namespace Opm {
+namespace DenseAd {
+
+/*!
+ * \brief Represents a function evaluation and its derivatives w.r.t. a fixed set of
+ *        variables.
+ */
+template <class ValueT>
+class EvaluationV
+{
+public:
+    //! field type
+    typedef ValueT ValueType;
+
+    //! number of derivatives
+    int size = 0;
+
+protected:
+    //! length of internal data vector
+    int length_ = 0;
+
+    //! position index for value
+    static constexpr int valuepos_ = 0;
+    //! start index for derivatives
+    static constexpr int dstart_ = 1;
+    //! end+1 index for derivatives
+    int dend_ = 0;
+
+public:
+    //! default constructor
+    // TODO: maybe we should make a default usable one?
+    EvaluationV() = default;
+
+    // TODO: this is rather dangerous to have
+    // put here just help to make ReturnEval_(EvaluationV, double) return EvaluationV.
+    // we should not use it in the calculation at all.
+    EvaluationV(const ValueType& value)
+     : size(0)
+     , length_(size + 1)
+     , dend_(length_)
+     , data_(length_, 0.0)
+    {
+        throw std::runtime_error(" You should not convert a Scalar to EvaluationV without the number of variables at all! ");
+        setValue(value);
+    }
+
+
+    //! when resizing, all the data will be cleaned
+    void resize(const int numVars) {
+        size = numVars;
+        length_ = size + 1;
+        dend_ = length_;
+        data_.resize(length_, 0.0);
+    }
+
+    //! copy other function evaluation
+    EvaluationV(const EvaluationV& other) = default;
+
+    // create an evaluation which represents a constant function
+    //
+    // i.e., f(x) = c. this implies an evaluation with the given value and all
+    // derivatives being zero.
+    template <class RhsValueType>
+    EvaluationV(const RhsValueType& c, const int numVars)
+    {
+        resize(numVars);
+        setValue( c );
+        // clearDerivatives();
+        Valgrind::CheckDefined( data_ );
+    }
+
+    // create an evaluation which represents a constant function
+    //
+    // i.e., f(x) = c. this implies an evaluation with the given value and all
+    // derivatives being zero.
+    template <class RhsValueType>
+    EvaluationV(const RhsValueType& c, const int varPos, const int numVars)
+    {
+        resize(numVars);
+
+        // The variable position must be in represented by the given variable descriptor
+        assert(0 <= varPos && varPos < size);
+
+        setValue( c );
+        // clearDerivatives();
```

Comment:
please add a comment before the call to resize() that this sets the derivatives to 0.